src/request/plan_builder.rs

```diff
@@ -0,0 +1,178 @@
+    /// to combine the responses from each region.
+    pub fn merge<In, M: Merge<In>>(self, merge: M) -> PlanBuilder<PdC, MergeResponse<P, In, M>, Ph>
+    where
+        In: Clone + Send + Sync + 'static,
+        P: Plan<Result = Vec<Result<In>>>,
+    {
+        PlanBuilder {
+            pd_client: self.pd_client.clone(),
+            plan: MergeResponse {
+                inner: self.plan,
+                merge,
+                phantom: PhantomData,
+            },
+            phantom: PhantomData,
+        }
+    }
+
+    /// Apply a processing step to a response (usually only needed if the request is sent to a
+    /// single region because post-porcessing can be incorporated in the merge step for multi-region
+    /// requests).
+    pub fn post_process(self) -> PlanBuilder<PdC, ProcessResponse<P, P::Result>, Ph>
+    where
+        P: Plan<Result: Process>,
+    {
+        PlanBuilder {
+            pd_client: self.pd_client.clone(),
+            plan: ProcessResponse {
+                inner: self.plan,
+                phantom: PhantomData,
+            },
+            phantom: PhantomData,
+        }
+    }
+}
+
+impl<PdC: PdClient, P: Plan + Shardable> PlanBuilder<PdC, P, NoTarget>
+where
+    P::Result: HasError,
+{
+    /// Split the request into shards sending a request to the region of each shard.
+    pub fn multi_region(self) -> PlanBuilder<PdC, MultiRegionPlan<P, PdC>, Targetted> {
+        PlanBuilder {
+            pd_client: self.pd_client.clone(),
+            plan: MultiRegionPlan {
+                inner: self.plan,
+                pd_client: self.pd_client,
+            },
+            phantom: PhantomData,
+        }
+    }
+}
+
+impl<PdC: PdClient, R: KvRequest + SingleKey> PlanBuilder<PdC, Dispatch<R>, NoTarget> {
+    /// Target the request at a single region.
+    pub async fn single_region(self) -> Result<PlanBuilder<PdC, Dispatch<R>, Targetted>> {
+        let key = self.plan.request.key();
+        let store = self.pd_client.clone().store_for_key(key.into()).await?;
+        set_single_region_store(self.plan, store, self.pd_client)
+    }
+}
+
+impl<PdC: PdClient, R: KvRequest> PlanBuilder<PdC, Dispatch<R>, NoTarget> {
+    /// Target the request at a single region; caller supplies the store to target.
+    pub async fn single_region_with_store(
+        self,
+        store: Store,
+    ) -> Result<PlanBuilder<PdC, Dispatch<R>, Targetted>> {
+        set_single_region_store(self.plan, store, self.pd_client)
+    }
+}
+
+fn set_single_region_store<PdC: PdClient, R: KvRequest>(
+    mut plan: Dispatch<R>,
+    store: Store,
+    pd_client: Arc<PdC>,
+) -> Result<PlanBuilder<PdC, Dispatch<R>, Targetted>> {
+    plan.request.set_context(store.region.context()?);
+    plan.kv_client = Some(store.client);
+    Ok(PlanBuilder {
+        plan,
+        pd_client,
+        phantom: PhantomData,
+    })
+}
+
+/// Indicates that a request operates on a single key.
+pub trait SingleKey {
```

Comment:
Who impls this trait?

Reply:
It is impl'ed for requests, e.g., `RawGetRequest: SingleKey`, it probably never needs to be impl'ed by users of the client, only implementors (i.e., we would need to add it for newly supported requests), but users of the crate might need to use it as a bound if they are doing something fancy.

Reply:
I still can't find any impl for requests by text-search. What did I miss?

Reply:
@andylokandy The impls are in raw/requests.rs and transaction/requests.rs. GitHub hides them 🤦🏻‍♂️